src/fu-util.c

```diff
@@ -673,6 +675,222 @@ fu_util_display_current_message(FuUtilPrivate *priv)
 	}
 }
 
+static gboolean
+fu_util_device_test_component(FuUtilPrivate *priv, JsonObject *json_obj, GBytes *fw, GError **error)
+{
+	JsonArray *json_array;
+	const gchar *name = "component";
+	const gchar *protocol = NULL;
+	g_autoptr(FwupdDevice) device = NULL;
+
+	/* some elements are optional */
+	if (json_object_has_member(json_obj, "name"))
+		name = json_object_get_string_member(json_obj, "name");
+	if (json_object_has_member(json_obj, "protocol"))
+		protocol = json_object_get_string_member(json_obj, "protocol");
+
+	/* find the device with any of the matching GUIDs */
+	if (!json_object_has_member(json_obj, "guids")) {
+		g_set_error_literal(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_INVALID_FILE,
+				    "JSON invalid as has no 'guids'");
+		return FALSE;
+	}
+	json_array = json_object_get_array_member(json_obj, "guids");
+	for (guint i = 0; i < json_array_get_length(json_array); i++) {
+		JsonNode *json_node = json_array_get_element(json_array, i);
+		FwupdDevice *device_tmp;
+		const gchar *guid = json_node_get_string(json_node);
+		g_autoptr(GPtrArray) devices = NULL;
+
+		g_debug("looking for guid %s", guid);
+		devices = fwupd_client_get_devices_by_guid(priv->client, guid, NULL, NULL);
+		if (devices == NULL)
+			continue;
+		if (devices->len > 1) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "multiple devices with GUID %s",
+				    guid);
+			return FALSE;
+		}
+		device_tmp = g_ptr_array_index(devices, 0);
+		if (protocol != NULL && !fu_device_has_protocol(device_tmp, protocol))
+			continue;
+		device = g_object_ref(device_tmp);
+		break;
+	}
+	if (device == NULL) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_NOT_SUPPORTED,
+			    "did not find any %s devices with matching GUIDs",
+			    name);
+		return FALSE;
+	}
+
+	/* verify the version matches what we expected */
+	if (json_object_has_member(json_obj, "version")) {
+		const gchar *version = json_object_get_string_member(json_obj, "version");
+		if (g_strcmp0(version, fu_device_get_version(device)) != 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "version of %s did not match, got %s, expected %s",
+				    name,
+				    fu_device_get_version(device),
+				    version);
+			return FALSE;
+		}
+	}
+
+	/* success */
+	return TRUE;
+}
+
+static gboolean
+fu_util_device_test_step(FuUtilPrivate *priv, JsonObject *json_obj, GError **error)
+{
+	JsonArray *json_array;
+	const gchar *basename;
+	g_autofree gchar *url = NULL;
+	g_autofree gchar *filename = NULL;
+	g_autoptr(GBytes) fw = NULL;
+
+	/* download file if required */
+	basename = json_object_get_string_member(json_obj, "file");
+	url = g_build_filename("https://fwupd.org/downloads", basename, NULL);
```

Comment:
rather than hardcode this in fwupdmgr, can it be an argument or config file option?  I would think fetching prefix from the remote configuration makes perfect sense for example.

Reply:
Maybe we just put the URL in the .JSON file? It doesn't have to be the LVFS, and in some cases it's better if it's not...

Reply:
Yeah, that's a better idea.

Reply:
Fixed.